core/drivers/crypto/caam/cipher/caam_cipher_mac.c

```diff
@@ -0,0 +1,859 @@
+		macdata->countdata += len;
+	}
+
+	return ret;
+}
+
+/*
+ * Finalize the MAC operation
+ *
+ * @ctx     Operation software context
+ * @digest  [out] Digest buffer
+ * @len     Digest buffer length
+ */
+static TEE_Result do_cmac_final(struct crypto_mac_ctx *ctx, uint8_t *digest,
+				size_t len)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	uint8_t *pad_src = NULL;
+	size_t pad_size = 0;
+	struct crypto_mac *mac = to_mac_ctx(ctx);
+	struct cipherdata *macdata = mac->ctx;
+	struct drvcrypt_cipher_update dupdate = { };
+
+	/* Check if the digest size is big enough */
+	if (len < macdata->alg->size_block)
+		return TEE_ERROR_SHORT_BUFFER;
+
+	if (macdata->mode == TEE_CHAIN_MODE_CBC_MAC_PKCS5) {
+		/* Calculate the last block PAD Size */
+		pad_size = macdata->alg->size_block;
+		pad_size -= macdata->countdata % macdata->alg->size_block;
+		CIPHER_TRACE("Pad size = %zu", pad_size);
+
+		if (pad_size) {
+			/* Need to pad the last block */
+			pad_src = caam_calloc(pad_size);
+
+			if (!pad_src) {
+				CIPHER_TRACE("Pad src allocation error");
+				return TEE_ERROR_OUT_OF_MEMORY;
+			}
+
+			memset(pad_src, pad_size, pad_size);
+		}
+	}
+
+	/* Prepare the update data */
+	dupdate.ctx = macdata;
+	dupdate.encrypt = true;
+	dupdate.last = true;
+	dupdate.src.data = pad_src;
+	dupdate.src.length = pad_size;
+	dupdate.dst.data = digest;
+	dupdate.dst.length = len;
+
+	ret = macdata->alg->update(&dupdate);
+
+	caam_free(pad_src);
+
+	return ret;
+}
+
+/*
+ * Free the software context
+ *
+ * @ctx    [in/out] Caller context variable
+ */
+static void do_cmac_free(struct crypto_mac_ctx *ctx)
+{
+	struct crypto_mac *mac = to_mac_ctx(ctx);
+
+	caam_cipher_free(mac->ctx);
+
+	free(mac);
+}
+
+/*
+ * Copy software cipher MAC context
+ *
+ * @dst_ctx  [out] Reference the context destination
+ * @src_ctx  Reference the context source
+ */
+static void do_cmac_copy_state(struct crypto_mac_ctx *dst_ctx,
+			       struct crypto_mac_ctx *src_ctx)
+{
+	struct crypto_mac *mac_src = to_mac_ctx(src_ctx);
+	struct crypto_mac *mac_dst = to_mac_ctx(dst_ctx);
+	struct cipherdata *macdata_dst = mac_dst->ctx;
+	struct cipherdata *macdata_src = mac_src->ctx;
+
+	caam_cipher_copy_state(macdata_dst, macdata_src);
+
+	macdata_dst->countdata = macdata_src->countdata;
+	macdata_dst->mode = macdata_src->mode;
+}
+
+/*
+ * Registration of the CMAC driver
+ */
+static const struct crypto_mac_ops cmac_ops = {
+	.init = do_cmac_init,
+	.update = do_cmac_update,
+	.final = do_cmac_final,
+	.free_ctx = do_cmac_free,
+	.copy_state = do_cmac_copy_state,
+};
+
+/*
+ * Allocate the software context
+ *
+ * @ctx      [out] Caller context variable
+ * @algo     Algorithm ID
+ */
+static TEE_Result caam_cmac_allocate(struct crypto_mac_ctx **ctx, uint32_t algo)
+{
+	TEE_Result ret = TEE_ERROR_NOT_IMPLEMENTED;
+	struct crypto_mac *mac = NULL;
+	const struct cipheralg *alg = NULL;
+	struct cipherdata *macdata = NULL;
+
+	CIPHER_TRACE("Allocate Context (%p) algo %" PRIx32, ctx, algo);
```

Comment:
suggest s/`algo %" PRIx32`/`algo %#PRIx32`/

Reply:
For consistency, I prefer to keep ```algo %" PRIx32```